drivers/resolver/ad2s1210/ad2s1210.c

```diff
@@ -0,0 +1,361 @@
+	if (ret)
+		return ret;
+
+	if (!desc)
+		return 0;
+
+	ret = no_os_gpio_direction_output(*desc, value);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+static void ad2s1210_gpio_remove(struct ad2s1210_dev *dev)
+{
+	if (dev->gpio_a0)
+		no_os_gpio_remove(dev->gpio_a0);
+
+	if (dev->gpio_a1)
+		no_os_gpio_remove(dev->gpio_a1);
+
+	if (dev->gpio_res0)
+		no_os_gpio_remove(dev->gpio_res0);
+
+	if (dev->gpio_res1)
+		no_os_gpio_remove(dev->gpio_res1);
+
+	if (dev->gpio_sample)
+		no_os_gpio_remove(dev->gpio_sample);
+}
+
+/*******************************************************************************
+ * @brief Initialize the ad2s1210 driver and create a descriptor.
+ *
+ * @param dev - Device descriptor to create.
+ * @param init_param - Initialization parameters.
+ *
+ * @return Returns negative error code or 0 in case of success.
+ *         Example: -EINVAL - Bad input parameters.
+ *                  -ENOMEM - Failed to allocate memory.
+ *                  0 - No errors encountered.
+*******************************************************************************/
+int32_t ad2s1210_init(struct ad2s1210_dev **dev,
+		      struct ad2s1210_init_param *init_param)
+{
+	struct ad2s1210_dev *d;
+	int32_t ret;
+
+
+	d = (struct ad2s1210_dev *)no_os_calloc(1, sizeof(*d));
+	if (!d)
+		return -ENOMEM;
+
+	d->name = ad2s1210_device_name[0];
+	d->mode = MODE_CONFIG;
+	d->resolution = init_param->resolution;
+
+	ret = ad2s1210_gpio_init(&d->gpio_a0, &init_param->gpio_a0,
+				 NO_OS_GPIO_HIGH, true);
+	if (ret)
+		goto err;
+
+	ret = ad2s1210_gpio_init(&d->gpio_a1, &init_param->gpio_a1,
+				 NO_OS_GPIO_HIGH, true);
+	if (ret)
+		goto err;
+
+	ret = ad2s1210_gpio_init(&d->gpio_res0, &init_param->gpio_res0,
+				 NO_OS_GPIO_LOW, true);
+	if (ret)
+		goto err;
+
+	ret = ad2s1210_gpio_init(&d->gpio_res1, &init_param->gpio_res1,
+				 NO_OS_GPIO_LOW, true);
+	if (ret)
+		goto err;
+
+	ret = ad2s1210_gpio_init(&d->gpio_sample, &init_param->gpio_sample,
+				 NO_OS_GPIO_LOW, false);
+	if (ret)
+		goto err;
+
+	ret = no_os_spi_init(&d->spi_desc, &init_param->spi_init);
+	if (ret)
+		goto err;
+
+	if (d->gpio_a0 && d->gpio_a1)
+		d->have_mode_pins = true;
+
+	if (d->gpio_res0 && d->gpio_res1)
+		d->have_resolution_pins = true;
+
+	ret = ad2s1210_set_resolution(d, d->resolution);
+	if (ret)
+		goto err_spi;
+
+	*dev = d;
+	return 0;
+
+err_spi:
+	no_os_spi_remove(d->spi_desc);
+err:
+	ad2s1210_gpio_remove(d);
+	no_os_free(d);
+	pr_err("%s initialization failed with status %ld\n", d->name, ret);
+
+	return ret;
+}
+
+int32_t ad2s1210_spi_single_conversion(struct ad2s1210_dev *dev,
+				       uint8_t chn,
+				       uint16_t *data)
+{
+	int32_t ret;
+	uint8_t addr;
+	uint8_t val[2] = {0};
+
+	if (chn > 1)
+		return -EINVAL;
+
+	ret = no_os_gpio_set_value(dev->gpio_sample, NO_OS_GPIO_LOW);
+	if (ret)
+		return ret;
+
+	ret = no_os_gpio_set_value(dev->gpio_sample, NO_OS_GPIO_HIGH);
+	if (ret)
+		return ret;
+
+	if (dev->have_mode_pins) {
+		if (chn == 0)
```

Comment:
`if (!ch)`

Reply:
chn is a int not a pointer.  should i still use ! instead of == 0?

Reply:
Yes, or even better:
```
if (chn)
     mode = MODE_VEL;
else
     mode = MODE_POS;
```

Reply:
ack